README.md
# dtparse

[![travisci](https://travis-ci.org/bspeice/dtparse.svg?branch=master)](https://travis-ci.org/bspeice/dtparse)
[![appveyor](https://ci.appveyor.com/api/projects/status/r4de76tg9utfjva1/branch/master?svg=true)](https://ci.appveyor.com/project/bspeice/dtparse/branch/master)
[![crates.io](https://img.shields.io/crates/v/dtparse.svg)](https://crates.io/crates/dtparse)
[![docs.rs](https://docs.rs/dtparse/badge.svg)](https://docs.rs/dtparse/)


The fully-featured "even I couldn't understand that" time parser.
Designed to take in strings and give back sensible dates and times.

dtparse has its foundations in the [`dateutil`](dateutil) library for
Python, which excels at taking "interesting" strings and trying to make
sense of the dates and times they contain. A couple of quick examples
from the test cases should give some context:

```rust
extern crate chrono;
extern crate dtparse;
use chrono::prelude::*;
use dtparse::parse;

assert_eq!(
    parse("2008.12.30"),
    Ok((NaiveDate::from_ymd(2008, 12, 30).and_hms(0, 0, 0), None))
);

// It can even handle timezones!
assert_eq!(
    parse("January 4, 2024; 18:30:04 +02:00"),
    Ok((
        NaiveDate::from_ymd(2024, 1, 4).and_hms(18, 30, 4),
        Some(FixedOffset::east(7200))
    ))
);
```

And we can even handle fuzzy strings where dates/times aren't the
only content if we dig into the implementation a bit!

```rust
extern crate chrono;
extern crate dtparse;
use chrono::prelude::*;
use dtparse::Parser;
use std::collections::HashMap;

let mut p = Parser::default();
assert_eq!(
    p.parse(
        "I first released this library on the 17th of June, 2018.",
        None, None,
        true /* turns on fuzzy mode */,
        true /* gives us the tokens that weren't recognized */,
        None, false, &HashMap::new()
    ),
    Ok((
        NaiveDate::from_ymd(2018, 6, 17).and_hms(0, 0, 0),
        None,
        Some(vec!["I first released this library on the ",
                  " of ", ", "].iter().map(|&s| s.into()).collect())
    ))
);
```

Further examples can be found in the [examples](examples) directory on international usage.

# Usage

`dtparse` requires a minimum Rust version of 1.21 to build, but is tested on Windows, OSX,
BSD, Linux, and WASM. The build is also compiled against the iOS and Android SDK's, but is not
tested against them.

[dateutil]: https://github.com/dateutil/dateutil
[examples]: https://github.com/bspeice/dtparse/tree/master/examples


CHANGELOG.md
Version 1.0.2 (2018-08-14)
==========================

Misc
----

- Add tests for WASM

Version 1.0.1 (2018-08-11)
==========================

Bugfixes
--------

- Fixed an issue with "GMT+3" not being handled correctly

Misc
----

- Upgrade `lazy_static` and `rust_decimal` dependencies

Version 1.0.0 (2018-08-03)
==========================

Initial release. Passes all relevant unit tests from Python's
`dateutil` project.


src/tests.rs
use ParseError;
use parse;

#[test]
fn test_fuzz() {

    assert_eq!(parse("\x2D\x38\x31\x39\x34\x38\x34"), Err(ParseError::InvalidMonth));
}

build_tests.py
#import dateutil.parser._timelex.split as time_split
from dateutil.parser import _timelex
from dateutil.parser import parse as duparse
import pytz

# The TEST_STRINGS list should be the only thing that actually needs changing
TEST_STRINGS = [
    '2018.5.15',
    'May 5, 2018',
    'Mar. 5, 2018',
    '19990101T23',
    '19990101T2359',
]

AUTOGEN_HEADER = '''
// WARNING
// This file was auto-generated using the `build_tests.py` script.
// Please do not edit it manually.

'''

S4 = ' ' * 4
S8 = ' ' * 8
S12 = ' ' * 12

def rust_tokenize(time_string):
    split_array = _timelex.split(time_string)
    return ['"{}".to_owned()'.format(token) for token in split_array]

def build_split_string_tests():
    header = '''use tokenize;

#[test]
fn test_python_compat() {\n'''

    tests = []

    for test_string in TEST_STRINGS:
        token_string = '\n'.join(['{}{},'.format(S12, s)
                                  for s in rust_tokenize(test_string)])
        tests.append('    assert_eq!(\n{}tokenize("{}"),\n{}vec![\n{}\n{}]\n{});'
                     .format(S8, test_string, S8, token_string, S8, S4))

    body = '\n'.join(tests)

    footer = '\n}\n'

    return header + body + footer

def test_parse(time_string):
    dt = duparse(time_string)
    # TODO: Don't make this dependent on New_York
    iso8601 = pytz.timezone('America/New_York').localize(dt).astimezone(pytz.utc)
    return 'assert_eq!(\n{}parse("{}".to_owned())\n{}.unwrap()\n{}.to_rfc3339_opts(SecondsFormat::Micros, false),\n{}"{}"\n{});'.format(
        S8, time_string, S12, S12, S8, iso8601, S4)

def build_parse_tests():
    header = '''use chrono::SecondsFormat;

use parse;

#[test]
fn test_python_compat() {\n'''

    asserts = ['    {}'.format(test_parse(a)) for a in TEST_STRINGS]
    body = '\n'.join(asserts)

    footer = '\n}\n'

    return header + body + footer

if __name__ == '__main__':
    split_string_test = build_split_string_tests()
    with open('src/tests/compat_split_string.rs', 'w+') as handle:
        handle.write(AUTOGEN_HEADER + split_string_test)

    parse_test = build_parse_tests()
    with open('src/tests/compat_parse.rs', 'w+') as handle:
        handle.write(AUTOGEN_HEADER + parse_test)

tests/parse.rs
extern crate chrono;
extern crate dtparse;

use chrono::DateTime;
use chrono::NaiveDate;
use chrono::NaiveDateTime;
use chrono::NaiveTime;
use chrono::Utc;

use dtparse::parse;

macro_rules! ymd_test {
    ($date:expr, $year:expr, $month:expr, $day:expr) => {
        let nd = NaiveDate::from_ymd($year, $month, $day);
        let nt = NaiveTime::from_hms(0, 0, 0);
        let dt = NaiveDateTime::new(nd, nt);
        let utc_dt = DateTime::from_utc(dt, Utc);

        let parsed = parse($date);

        println!("{:?}", parsed);
        assert!(parsed == Ok(utc_dt));
    };
}

#[test]
fn test_basic() {
    ymd_test!("2014 January 19", 2014, 1, 19);
}


ci/script.sh
# This script takes care of testing your crate

set -ex

main() {
    cross build --target $TARGET
    cross build --target $TARGET --release

    if [ ! -z $DISABLE_TESTS ]; then
        return
    fi

    cross test --target $TARGET
    cross test --target $TARGET --release
}

main_web() {
    CARGO_WEB_RELEASE="$(curl -L -s -H 'Accept: application/json' https://github.com/koute/cargo-web/releases/latest)"
    CARGO_WEB_VERSION="$(echo $CARGO_WEB_RELEASE | sed -e 's/.*"tag_name":"\([^"]*\)".*/\1/')"
    CARGO_WEB_URL="https://github.com/koute/cargo-web/releases/download/$CARGO_WEB_VERSION/cargo-web-x86_64-unknown-linux-gnu.gz"

    echo "Downloading cargo-web from: $CARGO_WEB_URL"
    curl -L "$CARGO_WEB_URL" | gzip -d > cargo-web
    chmod +x cargo-web

    mkdir -p ~/.cargo/bin
    mv cargo-web ~/.cargo/bin

    cargo web build --target $TARGET
    cargo web test --target $TARGET --release
}

# we don't run the "test phase" when doing deploys
if [ -z $TRAVIS_TAG ]; then
    if [ -z "$USE_CARGO_WEB" ]; then
        main
    else
        main_web
    fi
fi


src/weekday.rs
use ParseResult;
use ParseError;

#[derive(Debug, PartialEq)]
pub enum DayOfWeek {
    Sunday,
    Monday,
    Tuesday,
    Wednesday,
    Thursday,
    Friday,
    Saturday
}

impl DayOfWeek {

    pub fn to_numeral(&self) -> u32 {
        match *self {
            DayOfWeek::Sunday => 0,
            DayOfWeek::Monday => 1,
            DayOfWeek::Tuesday => 2,
            DayOfWeek::Wednesday => 3,
            DayOfWeek::Thursday => 4,
            DayOfWeek::Friday => 5,
            DayOfWeek::Saturday => 6,
        }
    }

    pub fn from_numeral(num: u32) -> DayOfWeek {
        match num % 7 {
            0 => DayOfWeek::Sunday,
            1 => DayOfWeek::Monday,
            2 => DayOfWeek::Tuesday,
            3 => DayOfWeek::Wednesday,
            4 => DayOfWeek::Thursday,
            5 => DayOfWeek::Friday,
            6 => DayOfWeek::Saturday,
            _ => panic!("Unreachable.")
        }
    }

    /// Given the current day of the week, how many days until the next day?
    pub fn difference(&self, other: &DayOfWeek) -> u32 {
        // Have to use i32 because of wraparound issues
        let s_num = self.to_numeral() as i32;
        let o_num = other.to_numeral() as i32;

        if o_num - s_num >= 0 {
            (o_num - s_num) as u32
        } else {
            (7 + o_num - s_num) as u32
        }
    }
}

pub fn day_of_week(year: u32, month: u32, day: u32) -> ParseResult<DayOfWeek> {
    // From https://en.wikipedia.org/wiki/Determination_of_the_day_of_the_week#Schwerdtfeger's_method
    let (c, g) = match month {
        3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 => {
            let c = year / 100;
            (c, year - 100 * c)
        },
        1 | 2 => {
            let c = (year - 1) / 100;
            (c, year - 1 - 100 * c)
        },
        _ => return Err(ParseError::ImpossibleTimestamp("Invalid month"))
    };

    let e = match month {
        1 | 5 => 0,
        2 | 6 => 3,
        3 | 11 => 2,
        4 | 7 => 5,
        8 => 1,
        9 | 12 => 4,
        10 => 6,
        _ => panic!("Unreachable.")
    };

    // This implementation is Gregorian-only.
    let f = match c % 4 {
        0 => 0,
        1 => 5,
        2 => 3,
        3 => 1,
        _ => panic!("Unreachable.")
    };

    match (day + e + f + g + g / 4) % 7 {
        0 => Ok(DayOfWeek::Sunday),
        1 => Ok(DayOfWeek::Monday),
        2 => Ok(DayOfWeek::Tuesday),
        3 => Ok(DayOfWeek::Wednesday),
        4 => Ok(DayOfWeek::Thursday),
        5 => Ok(DayOfWeek::Friday),
        6 => Ok(DayOfWeek::Saturday),
        _ => panic!("Unreachable.")
    }
}

// Rust warns about unused imports here, but they're definitely used.
#[allow(unused_imports)]
mod test {

    use weekday::day_of_week;
    use weekday::DayOfWeek;

    #[test]
    fn day_of_week_examples() {
        assert_eq!(day_of_week(2018, 6, 24).unwrap(), DayOfWeek::Sunday);
        assert_eq!(day_of_week(2003, 9, 25).unwrap(), DayOfWeek::Thursday);
    }

    #[test]
    fn weekday_difference() {

        assert_eq!(DayOfWeek::Sunday.difference(&DayOfWeek::Sunday), 0);
        assert_eq!(DayOfWeek::Sunday.difference(&DayOfWeek::Monday), 1);
        assert_eq!(DayOfWeek::Sunday.difference(&DayOfWeek::Tuesday), 2);
        assert_eq!(DayOfWeek::Sunday.difference(&DayOfWeek::Wednesday), 3);
        assert_eq!(DayOfWeek::Sunday.difference(&DayOfWeek::Thursday), 4);
        assert_eq!(DayOfWeek::Sunday.difference(&DayOfWeek::Friday), 5);
        assert_eq!(DayOfWeek::Sunday.difference(&DayOfWeek::Saturday), 6);
        assert_eq!(DayOfWeek::Monday.difference(&DayOfWeek::Sunday), 6);
        assert_eq!(DayOfWeek::Tuesday.difference(&DayOfWeek::Sunday), 5);
        assert_eq!(DayOfWeek::Wednesday.difference(&DayOfWeek::Sunday), 4);
        assert_eq!(DayOfWeek::Thursday.difference(&DayOfWeek::Sunday), 3);
        assert_eq!(DayOfWeek::Friday.difference(&DayOfWeek::Sunday), 2);
        assert_eq!(DayOfWeek::Saturday.difference(&DayOfWeek::Sunday), 1);
    }
}

examples/russian.rs
extern crate chrono;
extern crate dtparse;

use chrono::NaiveDate;
use dtparse::parse_info;
use dtparse::Parser;
use dtparse::ParserInfo;
use std::collections::HashMap;

fn main() {

    // In this example, we'll just swap the default "months" parameter
    // with a version in Russian. Lovingly taken from:
    // https://github.com/dateutil/dateutil/blob/99f5770e7c63aa049b28abe465d7f1cc25b63fd2/dateutil/test/test_parser.py#L244

    let mut info = ParserInfo::default();
    info.months = parse_info(vec![
        vec!["янв", "Январь"],
        vec!["фев", "Февраль"],
        vec!["мар", "Март"],
        vec!["апр", "Апрель"],
        vec!["май", "Май"],
        vec!["июн", "Июнь"],
        vec!["июл", "Июль"],
        vec!["авг", "Август"],
        vec!["сен", "Сентябрь"],
        vec!["окт", "Октябрь"],
        vec!["ноя", "Ноябрь"],
        vec!["дек", "Декабрь"]
    ]);

    let mut p = Parser::new(info);

    assert_eq!(
        p.parse("10 Сентябрь 2015 10:20", None, None, false, false, None, false, &HashMap::new())
            .unwrap().0,
        NaiveDate::from_ymd(2015, 9, 10).and_hms(10, 20, 0)
    );
}

build_pycompat_tokenizer.py
from dateutil.parser import _timelex

from build_pycompat import tests

def main():
    with open('src/tests/pycompat_tokenizer.rs', 'w+') as handle:
        handle.write(TEST_HEADER)

        counter = 0
        for _, test_strings in tests.items():
            for s in test_strings:
                handle.write(build_test(counter, s))
                counter += 1

def build_test(i, test_string):
    python_tokens = list(_timelex(test_string))
    formatted_tokens = 'vec!["' + '", "'.join(python_tokens) + '"]'
    return f'''
#[test]
fn test_tokenize{i}() {{
    let comp = {formatted_tokens};
    tokenize_assert("{test_string}", comp);
}}\n'''


TEST_HEADER = '''
//! This code has been generated by running the `build_pycompat_tokenizer.py` script
//! in the repository root. Please do not edit it, as your edits will be destroyed
//! upon re-running code generation.

use tokenize::Tokenizer;

fn tokenize_assert(test_str: &str, comparison: Vec<&str>) {
    let tokens: Vec<String> = Tokenizer::new(test_str).collect();
    assert_eq!(tokens, comparison, "Tokenizing mismatch for `{}`", test_str);
}\n'''

if __name__ == '__main__':
    main()

build_pycompat.py
from dateutil.parser import parse
from dateutil.tz import tzutc
from datetime import datetime

tests = {
    'test_parse_default': [
        "Thu Sep 25 10:36:28",
        "Sep 10:36:28", "10:36:28", "10:36", "Sep 2003", "Sep", "2003",
        "10h36m28.5s", "10h36m28s", "10h36m", "10h", "10 h 36", "10 h 36.5",
        "36 m 5", "36 m 5 s", "36 m 05", "36 m 05 s", "10h am", "10h pm",
        "10am", "10pm", "10:00 am", "10:00 pm", "10:00am", "10:00pm",
        "10:00a.m", "10:00p.m", "10:00a.m.", "10:00p.m.",
        "October", "31-Dec-00", "0:01:02", "12h 01m02s am", "12:08 PM",
        "01h02m03", "01h02", "01h02s", "01m02", "01m02h", "2004 10 Apr 11h30m",
        # testPertain
        'Sep 03', 'Sep of 03',
        # test_hmBY - Note: This appears to be Python 3 only, no idea why
        '02:17NOV2017',
        # Weekdays
        "Thu Sep 10:36:28", "Thu 10:36:28", "Wed", "Wednesday"
    ],
    'test_parse_simple': [
        "Thu Sep 25 10:36:28 2003", "Thu Sep 25 2003", "2003-09-25T10:49:41",
        "2003-09-25T10:49", "2003-09-25T10", "2003-09-25", "20030925T104941",
        "20030925T1049", "20030925T10", "20030925", "2003-09-25 10:49:41,502",
        "199709020908", "19970902090807", "2003-09-25", "09-25-2003",
        "25-09-2003", "10-09-2003", "10-09-03", "2003.09.25", "09.25.2003",
        "25.09.2003", "10.09.2003", "10.09.03", "2003/09/25", "09/25/2003",
        "25/09/2003", "10/09/2003", "10/09/03", "2003 09 25", "09 25 2003",
        "25 09 2003", "10 09 2003", "10 09 03", "25 09 03", "03 25 Sep",
        "25 03 Sep", "  July   4 ,  1976   12:01:02   am  ",
        "Wed, July 10, '96", "1996.July.10 AD 12:08 PM", "July 4, 1976",
        "7 4 1976", "4 jul 1976", "7-4-76", "19760704",
        "0:01:02 on July 4, 1976", "0:01:02 on July 4, 1976",
        "July 4, 1976 12:01:02 am", "Mon Jan  2 04:24:27 1995",
        "04.04.95 00:22", "Jan 1 1999 11:23:34.578", "950404 122212",
        "3rd of May 2001", "5th of March 2001", "1st of May 2003",
        '0099-01-01T00:00:00', '0031-01-01T00:00:00',
        "20080227T21:26:01.123456789", '13NOV2017', '0003-03-04',
        'December.0031.30',
        # testNoYearFirstNoDayFirst
        '090107',
        # test_mstridx
        '2015-15-May',
    ],
    'test_parse_tzinfo': [
        'Thu Sep 25 10:36:28 BRST 2003', '2003 10:36:28 BRST 25 Sep Thu',
    ],
    'test_parse_offset': [
        'Thu, 25 Sep 2003 10:49:41 -0300', '2003-09-25T10:49:41.5-03:00',
        '2003-09-25T10:49:41-03:00', '20030925T104941.5-0300',
        '20030925T104941-0300',
        # dtparse-specific
        "2018-08-10 10:00:00 UTC+3", "2018-08-10 03:36:47 PM GMT-4", "2018-08-10 04:15:00 AM Z-02:00"
    ],
    'test_parse_dayfirst': [
        '10-09-2003', '10.09.2003', '10/09/2003', '10 09 2003',
        # testDayFirst
        '090107',
        # testUnambiguousDayFirst
        '2015 09 25'
    ],
    'test_parse_yearfirst': [
        '10-09-03', '10.09.03', '10/09/03', '10 09 03',
        # testYearFirst
        '090107',
        # testUnambiguousYearFirst
        '2015 09 25'
    ],
    'test_parse_dfyf': [
        # testDayFirstYearFirst
        '090107',
        # testUnambiguousDayFirstYearFirst
        '2015 09 25'
    ],
    'test_unspecified_fallback': [
        'April 2009', 'Feb 2007', 'Feb 2008'
    ],
    'test_parse_ignoretz': [
        'Thu Sep 25 10:36:28 BRST 2003', '1996.07.10 AD at 15:08:56 PDT',
        'Tuesday, April 12, 1952 AD 3:30:42pm PST',
        'November 5, 1994, 8:15:30 am EST', '1994-11-05T08:15:30-05:00',
        '1994-11-05T08:15:30Z', '1976-07-04T00:01:02Z',
        'Tue Apr 4 00:22:12 PDT 1995'
    ],
    'test_fuzzy_tzinfo': [
        'Today is 25 of September of 2003, exactly at 10:49:41 with timezone -03:00.'
    ],
    'test_fuzzy_tokens_tzinfo': [
        'Today is 25 of September of 2003, exactly at 10:49:41 with timezone -03:00.'
    ],
    'test_fuzzy_simple': [
        'I have a meeting on March 1, 1974', # testFuzzyAMPMProblem
        'On June 8th, 2020, I am going to be the first man on Mars', # testFuzzyAMPMProblem
        'Meet me at the AM/PM on Sunset at 3:00 AM on December 3rd, 2003', # testFuzzyAMPMProblem
        'Meet me at 3:00 AM on December 3rd, 2003 at the AM/PM on Sunset', # testFuzzyAMPMProblem
        'Jan 29, 1945 14:45 AM I going to see you there?', # testFuzzyIgnoreAMPM
        '2017-07-17 06:15:', # test_idx_check
    ],
    'test_parse_default_ignore': [
    ],
}

def main():
    with open('src/tests/pycompat_parser.rs', 'w+') as handle:
        handle.write(TEST_HEADER)

        for test_name, test_strings in tests.items():
            for i, s in enumerate(test_strings):
                handle.write(globals()[test_name].__call__(i, s))


def test_parse_default(i, s):
    default = datetime(2003, 9, 25)
    d = parse(s, default=default)

    return TEST_PARSE_DEFAULT.format(i=i, d=d, s=s)


def test_parse_simple(i, s):
    d = parse(s)

    return TEST_PARSE_SIMPLE.format(i=i, d=d, s=s)


def test_parse_tzinfo(i, s):
    tzinfo = {'BRST': -10800}
    d = parse(s, tzinfos=tzinfo)

    return TEST_PARSE_TZINFO.format(i=i, d=d, s=s, offset=int(d.tzinfo._offset.total_seconds()))


def test_parse_offset(i, s):
    d = parse(s)
    return TEST_PARSE_OFFSET.format(i=i, d=d, s=s, offset=int(d.tzinfo._offset.total_seconds()))


def test_parse_dayfirst(i, s):
    d = parse(s, dayfirst=True)
    return TEST_PARSE_DAYFIRST.format(i=i, d=d, s=s)


def test_parse_yearfirst(i, s):
    d = parse(s, yearfirst=True)
    return TEST_PARSE_YEARFIRST.format(i=i, d=d, s=s)


def test_parse_dfyf(i, s):
    d = parse(s, dayfirst=True, yearfirst=True)
    return TEST_PARSE_DFYF.format(i=i, d=d, s=s)


def test_unspecified_fallback(i, s):
    d = parse(s, default=datetime(2010, 1, 31))
    return TEST_UNSPECIFIED_FALLBACK.format(i=i, d=d, s=s)


def test_parse_ignoretz(i, s):
    d = parse(s, ignoretz=True)
    return TEST_PARSE_IGNORETZ.format(i=i, d=d, s=s)


def test_parse_default_ignore(i, s):
    default = datetime(2003, 9, 25)
    d = parse(s, default=default)

    return TEST_PARSE_DEFAULT_IGNORE.format(i=i, d=d, s=s)


def test_fuzzy_tzinfo(i, s):
    d = parse(s, fuzzy=True)

    return TEST_FUZZY_TZINFO.format(i=i, d=d, s=s, offset=int(d.tzinfo._offset.total_seconds()))


def test_fuzzy_tokens_tzinfo(i, s):
    d, tokens = parse(s, fuzzy_with_tokens=True)

    r_tokens = ", ".join(list(map(lambda s: f'"{s}".to_owned()', tokens)))

    return TEST_FUZZY_TOKENS_TZINFO.format(
        i=i, d=d, s=s, offset=int(d.tzinfo._offset.total_seconds()),
        tokens=r_tokens
    )


def test_fuzzy_simple(i, s):
    d = parse(s, fuzzy=True)

    return TEST_FUZZY_SIMPLE.format(i=i, d=d, s=s)


# Here lies all the ugly junk.
TEST_HEADER = '''
//! This code has been generated by running the `build_pycompat.py` script
//! in the repository root. Please do not edit it, as your edits will be destroyed
//! upon re-running code generation.

extern crate chrono;

use chrono::Datelike;
use chrono::NaiveDate;
use chrono::NaiveDateTime;
use chrono::Timelike;
use std::collections::HashMap;

use Parser;
use ParserInfo;
use parse;

struct PyDateTime {
    year: i32,
    month: u32,
    day: u32,
    hour: u32,
    minute: u32,
    second: u32,
    micros: u32,
    tzo: Option<i32>
}

fn parse_and_assert(
    pdt: PyDateTime,
    info: ParserInfo,
    s: &str,
    dayfirst: Option<bool>,
    yearfirst: Option<bool>,
    fuzzy: bool,
    fuzzy_with_tokens: bool,
    default: Option<&NaiveDateTime>,
    ignoretz: bool,
    tzinfos: &HashMap<String, i32>,
) {

    let mut parser = Parser::new(info);
    let rs_parsed = parser.parse(
        s,
        dayfirst,
        yearfirst,
        fuzzy,
        fuzzy_with_tokens,
        default,
        ignoretz,
        tzinfos).expect(&format!("Unable to parse date in Rust '{}'", s));

    assert_eq!(pdt.year, rs_parsed.0.year(), "Year mismatch for '{}'", s);
    assert_eq!(pdt.month, rs_parsed.0.month(), "Month mismatch for '{}'", s);
    assert_eq!(pdt.day, rs_parsed.0.day(), "Day mismatch for '{}'", s);
    assert_eq!(pdt.hour, rs_parsed.0.hour(), "Hour mismatch for '{}'", s);
    assert_eq!(pdt.minute, rs_parsed.0.minute(), "Minute mismatch f'or' {}", s);
    assert_eq!(pdt.second, rs_parsed.0.second(), "Second mismatch for '{}'", s);
    assert_eq!(pdt.micros, rs_parsed.0.timestamp_subsec_micros(), "Microsecond mismatch for '{}'", s);
    assert_eq!(pdt.tzo, rs_parsed.1.map(|u| u.local_minus_utc()), "Timezone Offset mismatch for '{}'", s);
}

fn parse_and_assert_simple(
    pdt: PyDateTime,
    s: &str,
) {
    let rs_parsed = parse(s).expect(&format!("Unable to parse date in Rust '{}'", s));
    assert_eq!(pdt.year, rs_parsed.0.year(), "Year mismatch for '{}'", s);
    assert_eq!(pdt.month, rs_parsed.0.month(), "Month mismatch for '{}'", s);
    assert_eq!(pdt.day, rs_parsed.0.day(), "Day mismatch for '{}'", s);
    assert_eq!(pdt.hour, rs_parsed.0.hour(), "Hour mismatch for '{}'", s);
    assert_eq!(pdt.minute, rs_parsed.0.minute(), "Minute mismatch for '{}'", s);
    assert_eq!(pdt.second, rs_parsed.0.second(), "Second mismatch for '{}'", s);
    assert_eq!(pdt.micros, rs_parsed.0.timestamp_subsec_micros(), "Microsecond mismatch for '{}'", s);
    assert_eq!(pdt.tzo, rs_parsed.1.map(|u| u.local_minus_utc()), "Timezone Offset mismatch for '{}'", s);
}

fn parse_fuzzy_and_assert(
    pdt: PyDateTime,
    ptokens: Option<Vec<String>>,
    info: ParserInfo,
    s: &str,
    dayfirst: Option<bool>,
    yearfirst: Option<bool>,
    fuzzy: bool,
    fuzzy_with_tokens: bool,
    default: Option<&NaiveDateTime>,
    ignoretz: bool,
    tzinfos: &HashMap<String, i32>,
) {

    let mut parser = Parser::new(info);
    let rs_parsed = parser.parse(
        s,
        dayfirst,
        yearfirst,
        fuzzy,
        fuzzy_with_tokens,
        default,
        ignoretz,
        tzinfos).expect(&format!("Unable to parse date in Rust '{}'", s));

    assert_eq!(pdt.year, rs_parsed.0.year(), "Year mismatch for '{}'", s);
    assert_eq!(pdt.month, rs_parsed.0.month(), "Month mismatch for '{}'", s);
    assert_eq!(pdt.day, rs_parsed.0.day(), "Day mismatch for '{}'", s);
    assert_eq!(pdt.hour, rs_parsed.0.hour(), "Hour mismatch for '{}'", s);
    assert_eq!(pdt.minute, rs_parsed.0.minute(), "Minute mismatch f'or' {}", s);
    assert_eq!(pdt.second, rs_parsed.0.second(), "Second mismatch for '{}'", s);
    assert_eq!(pdt.micros, rs_parsed.0.timestamp_subsec_micros(), "Microsecond mismatch for '{}'", s);
    assert_eq!(pdt.tzo, rs_parsed.1.map(|u| u.local_minus_utc()), "Timezone Offset mismatch for '{}'", s);
    assert_eq!(ptokens, rs_parsed.2, "Tokens mismatch for '{}'", s);
}

macro_rules! rs_tzinfo_map {
    () => ({
        let mut h = HashMap::new();
        h.insert("BRST".to_owned(), -10800);
        h
    });
}\n'''

TEST_PARSE_DEFAULT = '''
#[test]
fn test_parse_default{i}() {{
    let info = ParserInfo::default();
    let default_rsdate = &NaiveDate::from_ymd(2003, 9, 25).and_hms(0, 0, 0);
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: None
    }};
    parse_and_assert(pdt, info, "{s}", None, None, false, false,
                     Some(default_rsdate), false, &HashMap::new());
}}\n'''

TEST_PARSE_SIMPLE = '''
#[test]
fn test_parse_simple{i}() {{
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: None,
    }};
    parse_and_assert_simple(pdt, "{s}");
}}\n'''

TEST_PARSE_TZINFO = '''
#[test]
fn test_parse_tzinfo{i}() {{
    let info = ParserInfo::default();
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: Some({offset}),
    }};
    parse_and_assert(pdt, info, "{s}", None, None, false, false,
                     None, false, &rs_tzinfo_map!());
}}\n'''

TEST_PARSE_OFFSET = '''
#[test]
fn test_parse_offset{i}() {{
    let info = ParserInfo::default();
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: Some({offset}),
    }};
    parse_and_assert(pdt, info, "{s}", None, None, false, false,
                     None, false, &HashMap::new());
}}\n'''

TEST_PARSE_DAYFIRST = '''
#[test]
fn test_parse_dayfirst{i}() {{
    let info = ParserInfo::default();
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: None,
    }};
    parse_and_assert(pdt, info, "{s}", Some(true), None, false, false,
                     None, false, &HashMap::new());
}}\n'''

TEST_PARSE_YEARFIRST = '''
#[test]
fn test_parse_yearfirst{i}() {{
    let info = ParserInfo::default();
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: None,
    }};
    parse_and_assert(pdt, info, "{s}", None, Some(true), false, false,
                     None, false, &HashMap::new());
}}\n'''

TEST_PARSE_DFYF = '''
#[test]
fn test_parse_dfyf{i}() {{
    let info = ParserInfo::default();
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: None,
    }};
    parse_and_assert(pdt, info, "{s}", Some(true), Some(true), false, false,
                     None, false, &HashMap::new());
}}\n'''

TEST_UNSPECIFIED_FALLBACK = '''
#[test]
fn test_unspecified_fallback{i}() {{
    let info = ParserInfo::default();
    let default_rsdate = &NaiveDate::from_ymd(2010, 1, 31).and_hms(0, 0, 0);
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: None
    }};
    parse_and_assert(pdt, info, "{s}", None, None, false, false,
                     Some(default_rsdate), false, &HashMap::new());
}}\n'''

TEST_PARSE_IGNORETZ = '''
#[test]
fn test_parse_ignoretz{i}() {{
    let info = ParserInfo::default();
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: None
    }};
    parse_and_assert(pdt, info, "{s}", None, None, false, false,
                     None, true, &HashMap::new());
}}\n'''

TEST_PARSE_DEFAULT_IGNORE = '''
#[test]
#[ignore]
fn test_parse_default_ignore{i}() {{
    let info = ParserInfo::default();
    let default_rsdate = &NaiveDate::from_ymd(2003, 9, 25).and_hms(0, 0, 0);
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: None
    }};
    parse_and_assert(pdt, info, "{s}", None, None, false, false,
                     Some(default_rsdate), false, &HashMap::new());
}}\n'''

TEST_FUZZY_TZINFO = '''
#[test]
fn test_fuzzy_tzinfo{i}() {{
    let info = ParserInfo::default();
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: Some({offset})
    }};
    parse_fuzzy_and_assert(pdt, None, info, "{s}", None, None, true, false,
                           None, false, &HashMap::new());
}}\n'''

TEST_FUZZY_TOKENS_TZINFO = '''
#[test]
fn test_fuzzy_tokens_tzinfo{i}() {{
    let info = ParserInfo::default();
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: Some({offset})
    }};
    let tokens = vec![{tokens}];
    parse_fuzzy_and_assert(pdt, Some(tokens), info, "{s}", None, None, true, true,
                           None, false, &HashMap::new());
}}\n'''

TEST_FUZZY_SIMPLE = '''
#[test]
fn test_fuzzy_simple{i}() {{
    let info = ParserInfo::default();
    let pdt = PyDateTime {{
        year: {d.year}, month: {d.month}, day: {d.day},
        hour: {d.hour}, minute: {d.minute}, second: {d.second},
        micros: {d.microsecond}, tzo: None
    }};
    parse_fuzzy_and_assert(pdt, None, info, "{s}", None, None, true, false,
                           None, false, &HashMap::new());
}}\n'''


if __name__ == '__main__':
    main()